src/cpg_flow/workflow.py

```diff
@@ -425,54 +391,39 @@ def set_stages(
         logger.info(f'  workflow/first_stages: {first_stages}')
         logger.info(f'  workflow/last_stages: {last_stages}')
 
-        # Round 1: initialising stage objects.
+        # Iinitialise stage objects.
         _stages_d: dict[str, Stage] = {}
+
+        def _make_once(cls) -> tuple['Stage', bool]:
+            try:
+                return _stages_d[cls.__name__], False
+            except KeyError:
+                instance = _stages_d[cls.__name__] = cls()
+                return instance, True
+
+        def _recursively_make_stage(cls):
+            instance, is_new = _make_once(cls)
+            if is_new:
+                instance.skipped = cls.__name__ in skip_stages
+                if not instance.skipped:
+                    instance.required_stages.extend(
+                        filter(None, map(_recursively_make_stage, instance.required_stages_classes)),
+                    )
+            return instance
+
         for cls in requested_stages:
-            if cls.__name__ in _stages_d:
-                continue
-            _stages_d[cls.__name__] = cls()
-
-        # Round 2: depth search to find implicit stages.
-        depth = 0
-        while True:  # might require few iterations to resolve dependencies recursively
-            depth += 1
-            newly_implicitly_added_d = dict()
-            for stg in _stages_d.values():
-                if stg.name in skip_stages:
-                    stg.skipped = True
-                    continue  # not searching deeper
-
-                if only_stages and stg.name not in only_stages:
-                    stg.skipped = True
-
-                # Iterate dependencies:
-                for reqcls in stg.required_stages_classes:
-                    if reqcls.__name__ in _stages_d:  # already added
-                        continue
-                    # Initialising and adding as explicit.
-                    reqstg = reqcls()
-                    newly_implicitly_added_d[reqstg.name] = reqstg
-
-            if newly_implicitly_added_d:
-                logger.info(
-                    f'Additional implicit stages: {list(newly_implicitly_added_d.keys())}',
-                )
-                _stages_d |= newly_implicitly_added_d
-            else:
-                # No new implicit stages added, so can stop the depth-search here
-                break
+            _recursively_make_stage(cls)
 
-        # Round 3: set "stage.required_stages" fields to each stage.
-        for stg in _stages_d.values():
-            stg.required_stages = [
-                _stages_d[cls.__name__] for cls in stg.required_stages_classes if cls.__name__ in _stages_d
-            ]
+        implicit_stages = set(_stages_d.keys()) - {stage.__name__ for stage in requested_stages}
+        logger.info(
+            f'Additional implicit stages: {sorted(implicit_stages)}',
```

Comment:
As a UX note, it would be good to clarify the sort here. If this sort is on the String names of Stages, `sorted` is going to produce a random ordering relative to the true order, so making it less user friendly to interpret.

Looking back at the stage discovery it's probably a bit of an ordeal to order this list of Strings.

Potentially we could grow a list of encountered Stage names in `_recursively_make_stage`, flip the ordering, and use that as a start -> finish order for this logging message.